File: solutions/day22_part1.ts
import fs from "node:fs";

const raw = fs.readFileSync("./input.txt", "utf-8");
const numbers = raw.split("\n").map(Number);

function generateNewNumber(number: number): number {
  const mix = (value: number, secret: number): number => {
    return (value ^ secret) >>> 0;
  };
  const prune = (secret: number): number => {
    return secret % 16777216;
  };
  const step1 = prune(mix(number * 64, number));
  const step2 = prune(mix(Math.floor(step1 / 32), step1));
  const step3 = prune(mix(step2 * 2048, step2));
  return step3;
}

const result = numbers.reduce<number>((acc, number) => {
  const iterate = (remaining: number, iteratedNumber: number) => {
    if (remaining === 0) return iteratedNumber;
    return iterate(remaining - 1, generateNewNumber(iteratedNumber));
  };
  acc += iterate(2000, number);
  return acc;
}, 0);
console.log(result);

File: solutions/python/day22_part1.py
with open("./input.txt", "r") as txt:
  num: str = txt.read().rstrip()
  txt.close()

numbers = list(map(int, num.split('\n')))

def mix(value: int, secret: int) -> int:
  return value ^ secret
def prune(secret: int) -> int:
  return secret % 16777216
def generateNewNumber(number: int) -> int:
  step1 = prune(mix(number * 64, number))
  step2 = prune(mix(step1 // 32, step1))
  step3 = prune(mix(step2 * 2048, step2))
  return step3

for i in range(2000):
  for j in range(len(numbers)):
    numbers[j] = generateNewNumber(numbers[j])

result = 0
for number in numbers:
  result += number
print(result)
